### ISSUE TITLE
CSP

### ISSUE THREAD

**Issue by cdruet** [miguelgrinberg/Flask-Moment#43]
I added Talisman to my Flask project and tried to properly configure TALISMAN_SECURITY_POLICY but I can't figure out how to get Moment scripts accepted.

I got the following message(s): _Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self' 'unsafe-inline' cdnjs.cloudflare.com maxcdn.bootstrapcdn.com code.jquery.com 'nonce-gOu5h3Vkvkid_uTOnxjYsw'". Note that 'unsafe-inline' is ignored if either a hash or nonce value is present in the source list._

I tried to investigate things a bit further and it looks like it comes from `moment.include_jquery()` and `moment.include_moment()`. I wonder if someone (@miguelgrinberg ?) could add an option to include a SHA or a nonce to the script tag (if that's the proper way to do it) or let me know how to circumvent the issue(s).

**Comment by miguelgrinberg** [comment]
You can call `moment.include_moment(no_js=True)` and then it is your own responsibility to load moment.js, so you can add the hash.

**Comment by malan88** [comment]
@miguelgrinberg I have disabled the CSP for the same reason, but I found that one of the key issues with using nonce and flask-moment is that I cannot specify a nonce per script attribute.

The `moment.include_jquery()` and `moment.include_moment()` problems are easy to solve: adding `cdnjs.cloudflare.com` and `code.jquery.com` to the [allowable `script-src` works perfectly for that](https://github.com/GoogleCloudPlatform/flask-talisman#example-6). But in order to run inline-script that is generated by a flask moment object I need to be able to specify a nonce attribute in the script tag:
```
<script nonce="{{ csp_nonce() }}">
    //...
</script>
```
This is not an issue with a lot of generated jinja2, for instance with Flask-WTF, because it seems I can render arbitrary (or semi-arbitrary) html attributes when inserting those pieces of jinja2, but flask moment doesn't seem to be capable of rendering such attributes.

If there was a way to either (a) add arbitrary html attributes to the script tag when creating and rendering a flask moment object or (b) a way to render *just the javascript* and surround it with script tags myself this would not be an issue.

Miguel, your software and tutorials are incredible. I am hoping you can help us on this one. Thanks.

**Comment by miguelgrinberg** [comment]
@malan88 why can't you just copy the JS and include it manually in your page? Don't use the `include_moment()` function if it doesn't do what you want.

**Comment by malan88** [comment]
Miguel, I seem to have misunderstood the nature of the code (and what you were saying about it) but can now see another issue at present.

I now realize that `include_moment()` renders *both* the link to the moment.js cdn *and* a snippet of jquery that you yourself wrote. I have successfully copied this snippet and added the nonce and now it renders but there is still further a problem. Specifically, you use `eval` in the snippet and this is by default disallowed by CSP. It can be relaxed by adding `unsafe-eval` to the definition of a CSP but is [strongly recommended against][0].

Let me ask several questions: does Flask-Moment use JQuery anywhere other than that snippet of JS that you wrote? And would you mind me attempting to write a PR that refactors your JS to no longer use `eval` *and* use vanilla JavaScript? I say this because in my own work I avoid JQuery and explicitly write vanilla JS and the only need I have for JQuery is Flask-Moment. If I attempt to refactor Flask-Moment to not use eval I feel like it would benefit the library to also eliminate an added JS dependency. If I were to rewrite this snippet in vanilla JS, would the JQuery requirement become superfluous?

I will mention that this might take me a bit of time (probably a week, but perhaps today, not sure) because it would be my first code PR.

Thanks, Miguel.

[0]: https://developer.chrome.com/extensions/contentSecurityPolicy#relaxing-eval

**Comment by miguelgrinberg** [comment]
@malan88 I think removing eval is going to be harder than you estimate, but as a high level concept I agree that it would be better not to use it, so I would happily accept a PR that does it as long as it does not break existing uses and is relatively simple (i.e. not a complete re-implementation of eval without using eval).

Removing jQuery as a dependency is also fine, I actually wanted to do that myself but never got around to do it.

**Comment by malan88** [comment]
@miguelgrinberg 

You are right that it will be harder than I initially suspected. But I think I can do it all in js without touching the actual python.

I will be working on it today and a buddy who is better at JS than me will be giving me a hand.

Thanks.

**Comment by malan88** [comment]
@miguelgrinberg 

My strategy in this instance is (was) to write a dictionary that decomposes the function strings from data-format into a set of arguments and then translate them using an object into actual function calls *which would eliminate the ability for people to arbitrarily execute code*, but this *is* essentially what you are saying about "reimplementing eval without eval" because I'm basically just writing a parser/translator. So it seems this is not what you would like me to do.

The other strategy would be to modify the way you render the spans, adding several `dataset` attributes, and then making lambdas that call those datasets in an object.

This actually seems cleaner, but is obviously more invasive.

I would like your opinion before I proceed.

**Comment by miguelgrinberg** [comment]
> which would eliminate the ability for people to arbitrarily execute code

I believe it is currently not possible for people to arbitrarily execute code. What goes in the eval is not provided by the user, it is code that is generated by this package and it is always a call to `moment()` with varying arguments that depend on the format requested. I don't believe there is a way to make this code do anything other than that, but if you've found a vulnerability I'd like to know about it.

As far as how to remove the eval, that is basically my problem, it is actually non-trivial. I think both of your options are sort of reimplementing the eval functionality either by parsing a single string, or a bunch of data attributes. I'm honestly not sure it is even worth the trouble, because this extension doesn't really do much, it would be fairly easy for an application to implement the same thing directly between Jinja and moment.js, and then you would not need to be generic so the eval would not be necessary.

Out of the two options I think the second is more reasonable, but it seems fairly complex still. Another option that may end up being simpler is to generate JavaScript code from the Python side, but I'm not sure if this is feasible, I didn't really think about it much.

**Comment by malan88** [comment]
So I actually went ahead and wrote it. It is not perfect and I think that it can be improved. I'm going to open a PR. Let's discuss it more there.

**Comment by malan88** [comment]
So I just wanted to note that it seems the solution to this is as follows:

Set the CSP's script-src to include both 'code.jquery.com' and 'cdnjs.cloudflare.com'. Also include "'unsafe-eval'" and set content_security_policy_nonce_in.

```
csp = {
   'default-src': "'self'",
   'script-src': ['code.jquery.com', 'cdnjs.cloudflare.com', "'self'", "'unsafe-evail'"]
}
talisman = Talisman(
    app,
    content_security_policy=csp,
    content_security_policy_nonce_in=['script-src']
)
```

Then, instead of using `include_moment()` in the template, copy and paste the output of it directly into the template and make sure that the jquery snippet has the following in the opening tag:

`<script nonce="{{ csp_nonce() }}">`

That should solve the problem.

**Comment by e2jk** [comment]
Hi @cdruet and @malan88 (and obviously @miguelgrinberg), in case you're still interested in removing that `unsafe-eval` from your CSP, please have a look at my proposed PR #61.
Have a great day!

**Comment by e2jk** [comment]
FYI @cdruet and @malan88, this has been fixed in the latest release 0.10.0, available now on PyPI. Thanks to Miguel for taking the time to fix this.

**Comment by cdruet** [comment]
> FYI @cdruet and @malan88, this has been fixed in the latest release 0.10.0, available now on PyPI. Thanks to Miguel for taking the time to fix this.

Still very much interested. Huge thanks to @miguelgrinberg !!!

**Comment by miguelgrinberg** [comment]
@e2jk you are selling yourself short. You actually wrote this fix, I just touched a couple things and merged. Thanks should go to you!
